File: 0011-container-with-most-water/0011-container-with-most-water.cpp
class Solution {
public:
    int maxArea(vector<int>& height) {
        /*
        we can solve this using two pointer approach
        anytime the water volume of water will be
        water = min(height[l], height[r]) * (r-l)
        we need to maximize the above thing...
        we can be a little greedy by moving the pointer which has the min height cause moving the bigger doesn't make sense cause the width (r-l) is gonna decreased anyway so it makes sense to move the smaller one inward.

        */
        int l = 0, r = height.size()-1, bestArea = 0;
        while(l<r){
            int area = min(height[l], height[r]) * (r-l);
            bestArea = max(area, bestArea);
            if(height[l]<height[r]) l++;
            else r--;
        }
        return bestArea;
    }
};

File: 0011-container-with-most-water/0011-container-with-most-water.go
func maxArea(height []int) int {
    l, r := 0, len(height)-1
    maxA := 0
    for l<r{
        maxA = max(maxA, (r-l)*min(height[l], height[r]))
        if height[l]>height[r]{
            r--
        } else {
            l++
        }
    }
    return maxA
}